feat(app): open graph image generator added

hanifisenturk opened the pull request:
Hi @armagandalkiran, @aahmetcakir !

I added the meta tag for previewing links. I also created an endpoint that generates og images by using edge runtime.

At the moment it is not necessary but we can use that endpoint to show the name of the page which we will visit in the future by using params.

And this is the demo preview of my localhost:
<img width="1512" alt="image" src="https://github.com/Trendyol/gurubu/assets/65610514/86cb42e8-114d-4eb1-83b1-785ccc82bafc">

hanifisenturk commented:
This PR has been created for #108

mertcanaltin commented on gurubu-client/src/app/api/og/route.tsx:
```diff
@@ -0,0 +1,55 @@
+    return new ImageResponse(
+      (
+        <div
+          style={{
+            backgroundColor: "#ffffff",
+            backgroundSize: "150px 150px",
+            height: "100%",
+            width: "100%",
+            display: "flex",
+            textAlign: "center",
+            alignItems: "center",
+            justifyContent: "center",
+            flexDirection: "column",
+            flexWrap: "nowrap",
+            fontSize: "48px",
```
it may not be healthy to keep this inline css

mertcanaltin replied:
Hmm I see, I hope it won't be crowded here 🤔 😄

mertcanaltin commented on gurubu-client/src/app/api/og/route.tsx:
```diff
@@ -0,0 +1,55 @@
+            fontSize: "48px",
+          }}
+        >
+          <svg
+            width="496"
+            height="120"
+            viewBox="0 0 124 30"
+            fill="none"
+            xmlns="http://www.w3.org/2000/svg"
+          >
+            <path
+              d="M17.3424 16.3333C16.9852 12.227 13.3578 9 8.93564 9C4.27677 9 0.5 12.5817 0.5 17C0.5 21.4183 4.27677 25 8.93564 25H21.5891V16.3333H17.3424Z"
+              fill="#528BFF"
+            />
+            <ellipse cx="20.8862" cy="15" rx="10.5446" ry="10" fill="#7839EE" />
+            <path
+              d="M42.9747 22.752C42.1107 22.752 41.3247 22.56 40.6167 22.176C39.9207 21.78 39.3627 21.21 38.9427 20.466C38.5347 19.71 38.3307 18.792 38.3307 17.712V14.688C38.3307 13.068 38.7867 11.826 39.6987 10.962C40.6107 10.086 41.8467 9.648 43.4067 9.648C44.9547 9.648 46.1487 10.062 46.9887 10.89C47.8407 11.706 48.2667 12.816 48.2667 14.22V14.292H45.9267V14.148C45.9267 13.704 45.8307 13.302 4 ...
+              fill="black"
+            />
+          </svg>
```
this svg is pretty big, let's import it from somewhere

hanifisenturk replied:
I exported the logo with text as png format, I removed this whole messy svg data :).

mertcanaltin commented on gurubu-client/src/app/api/og/route.tsx:
```diff
@@ -0,0 +1,55 @@
+    );
+  } catch (e: any) {
+    console.error(`${e.message}`);
+    return new Response(`Failed to generate the image`, {
+      status: 500,
```
let's discuss here

Is it healthy to return the user directly to 500?

hanifisenturk replied:
I agree btw @mertcanaltin , so I changed this a little bit. When something fails, I also return an ImageResponse but without png data, only text.

Normal condition:
<img width="1512" alt="Screenshot 2024-02-18 at 12 07 38" src="https://github.com/Trendyol/gurubu/assets/65610514/37f18c76-5eaa-4178-ad0c-27803c3601a3">

In error condition:
<img width="1512" alt="Screenshot 2024-02-18 at 12 07 21" src="https://github.com/Trendyol/gurubu/assets/65610514/f7762cc0-7fa2-48a3-bd84-59d211ca767a">

mertcanaltin replied:
@armagandalkiran @aahmetcakir @hanifisenturk Should we show the cause of the error here?

mertcanaltin commented on gurubu-client/src/app/api/og/route.tsx:
```diff
@@ -0,0 +1,53 @@
+      (
+        <div style={imageStyle}>
+          {imageData && (
+            <img width={372} height={90} src={imageData as unknown as string} />
```
It would be great if we could make this place cleaner. 
```js
src={imageData as unknown as string}
```

mertcanaltin commented on gurubu-client/src/app/api/og/route.tsx:
```diff
@@ -0,0 +1,53 @@
+
+export async function GET(request: Request) {
+  try {
+    const imageData = await fetch(
+      new URL("../../../../public/logo-with-text.png", import.meta.url)
+    ).then((res) => res.arrayBuffer());
```
is it possible to take it as an ordinary picture?

hanifisenturk replied:
@mertcanaltin actually I thought that actually. As I used the edge runtime for this operation, all resources have to get with fetch api. [This link](https://vercel.com/docs/functions/og-image-generation#edge-runtime) explains that as well. Of course, if I used Node runtime, we can get the data via `fs.readFile`.

mertcanaltin commented:
everything looks great. Thank you @hanifisenturk 🚀 
I left a few small comments

@armagandalkiran and @aahmetcakir  will check at the appropriate time 🎸

hanifisenturk commented:
Hi! @armagandalkiran @aahmetcakir @mertcanaltin , it's me, again!  😄

I updated the og logic, removed the API endpoint. At the moment, we are using NextJS's built-in og feature. It  generates necessary meta tags for us.

mertcanaltin commented:
Greetings @hanifisenturk , first of all sorry for the late reply, thank you very much for your efforts, I will be reviewing this pr in the future, everything looks great LGTM ;)

mertcanaltin commented on gurubu-client/src/app/opengraph-image.tsx:
```diff
@@ -0,0 +1,42 @@
+        <img
+          width={372}
+          height={90}
+          src={"https://gurubu.vercel.app/gurubu-logo.svg"}
```
```suggestion
          src={"/gurubu-logo.svg"}
```
